internal/auth/repository/eventsourcing/eventstore/key.go

```diff
@@ -40,17 +71,98 @@ func (k *KeyRepository) GetKeySet(ctx context.Context) (*jose.JSONWebKeySet, err
+	}
+	timer = time.After(duration)
+}
+
+func (k *KeyRepository) latestSigningKey() (shortRefresh *model.KeyView, err error) {
+	key, errView := k.View.GetActivePrivateKeyForSigning(time.Now().UTC().Add(k.SigningKeyGracefulPeriod))
+	if errView != nil && !errors.IsNotFound(errView) {
+		logging.Log("EVENT-GEd4h").WithError(errView).Warn("could not get signing key")
+		return nil, errView
+	}
+	return key, nil
+}
+
+func (k *KeyRepository) ensureIsLatestKey(ctx context.Context) (bool, error) {
+	sequence, err := k.View.GetLatestKeySequence()
 	if err != nil {
-		errCh <- err
-		return
+		return false, err
+	}
+	events, err := k.getKeyEvents(ctx, sequence.CurrentSequence)
+	if err != nil {
+		logging.Log("EVENT-der5g").WithError(err).Warn("error retrieving new events")
+		return false, err
+	}
+	if len(events) > 0 {
+		logging.Log("EVENT-GBD23").Warn("view not up to date, retrying later")
+		return false, nil
+	}
+	return true, nil
+}
+
+func (k *KeyRepository) refreshSigningKey(ctx context.Context, key *model.KeyView, keyCh chan<- jose.SigningKey, algorithm string) (refreshed bool, err error) {
+	if key == nil {
+		if k.currentKeyExpiration.Before(time.Now().UTC()) {
+			keyCh <- jose.SigningKey{}
+		}
+		if ok, err := k.ensureIsLatestKey(ctx); !ok && err == nil {
+			return false, err
+		}
+		err = k.lockAndGenerateSigningKeyPair(ctx, algorithm)
+		logging.Log("EVENT-B4d21").OnError(err).Warn("could not create signing key")
+		return false, err
 	}
+	if k.currentKeyID == key.ID {
+		return false, nil
+	}
+	if ok, err := k.ensureIsLatestKey(ctx); !ok && err == nil {
+		return false, err
+	}
+	signingKey, err := model.SigningKeyFromKeyView(key, k.KeyAlgorithm)
+	if err != nil {
+		return false, err
+	}
+	k.currentKeyID = signingKey.ID
+	k.currentKeyExpiration = key.Expiry
 	keyCh <- jose.SigningKey{
-		Algorithm: jose.SignatureAlgorithm(key.Algorithm),
+		Algorithm: jose.SignatureAlgorithm(signingKey.Algorithm),
 		Key: jose.JSONWebKey{
-			KeyID: key.ID,
-			Key:   key.Key,
+			KeyID: signingKey.ID,
+			Key:   signingKey.Key,
 		},
 	}
+	return true, nil
+}
+
+func (k *KeyRepository) lockAndGenerateSigningKeyPair(ctx context.Context, algorithm string) error {
+	err := k.Locker.Renew(k.lockerID(), signingKey, k.SigningKeyRotationCheck*2)
+	if err != nil {
+		if errors.IsErrorAlreadyExists(err) {
+			return nil
+		}
+		return err
+	}
+	return k.Commands.GenerateSigningKeyPair(ctx, algorithm)
+}
+
+func (k *KeyRepository) lockerID() string {
+	if k.lockID == "" {
+		var err error
+		k.lockID, err = os.Hostname()
+		if err != nil || k.lockID == "" {
+			k.lockID, err = id.SonyFlakeGenerator.Next()
+			logging.Log("EVENT-bsdf6").OnError(err).Panic("unable to generate lockID")
+		}
+		context.TODO()
+	}
+	return k.lockID
+}
+
+func (k *KeyRepository) getKeyEvents(ctx context.Context, sequence uint64) ([]eventstore.EventReader, error) {
+	return k.Eventstore.FilterEvents(ctx, key_view.KeyPairQuery(sequence))
```

Comment:
did you use eventstore 2 on purpose?

Reply:
which else should i use? v1 will be removed / deprecated when new-eventstore will be merged into master